components/core/tests/test-ffi_utils.cpp

```diff
@@ -0,0 +1,197 @@
+            constexpr int8_t cHeaderSignature{static_cast<int8_t>(0xC0)};  // 0b1100_0000
+            auto const num_bits_shift{num_continuation_bytes - 1};
+            auto const header_payload_mask{
+                    static_cast<uint8_t>(cHeaderPayloadMask >> num_bits_shift)
+            };
+            auto const header_signature{static_cast<uint8_t>(cHeaderSignature >> num_bits_shift)};
+            encoded_bytes.push_back(static_cast<char>(
+                    (least_significant_byte & header_payload_mask) | header_signature
+            ));
+            break;
+        }
+    }
+    return {encoded_bytes.rbegin(), encoded_bytes.rend()};
+}
+}  // namespace
+
+TEST_CASE("escape_utf8_string_basic", "[ffi][utils]") {
+    std::string test_str;
+    std::optional<std::string> actual;
+
+    // Test empty string
+    actual = validate_and_escape_utf8_string(test_str);
+    REQUIRE((actual.has_value() && actual.value() == get_expected_escaped_string(test_str)));
+
+    // Test string that has nothing to escape
+    test_str = "This string has nothing to escape :)";
+    actual = validate_and_escape_utf8_string(test_str);
+    REQUIRE((actual.has_value() && actual.value() == get_expected_escaped_string(test_str)));
+
+    // Test string with all single byte UTF-8 characters, including those we escape.
+    test_str.clear();
+    for (uint8_t i{0}; i <= static_cast<uint8_t>(INT8_MAX); ++i) {
+        test_str.push_back(static_cast<char>(i));
+    }
+    // Shuffle characters randomly
+    // NOLINTNEXTLINE(cert-msc32-c, cert-msc51-cpp)
+    std::shuffle(test_str.begin(), test_str.end(), std::default_random_engine{});
+    actual = validate_and_escape_utf8_string(test_str);
+    REQUIRE((actual.has_value() && actual.value() == get_expected_escaped_string(test_str)));
+
+    // Test valid UTF8 chars with continuation bytes
+    std::vector<std::string> const valid_utf8{
+            "\n",
+            "\xF0\xA0\x80\x8F",  // https://en.wiktionary.org/wiki/%F0%A0%80%8F
+            "a",
+            "\xE4\xB8\xAD",  // https://en.wiktionary.org/wiki/%E4%B8%AD
+            "\x1F",
+            "\xC2\xA2",  // ¢
+            "\\"
+    };
+    test_str.clear();
+    for (auto const& str : valid_utf8) {
+        test_str.append(str);
+    }
+    actual = validate_and_escape_utf8_string(test_str);
+    REQUIRE((actual.has_value() && actual.value() == get_expected_escaped_string(test_str)));
+}
+
+TEST_CASE("escape_utf8_string_with_invalid_continuation", "[ffi][utils]") {
+    std::string test_str;
+
+    auto const valid_utf8_byte_sequence = GENERATE(
+            generate_utf8_byte_sequence(0x80, 1),
+            generate_utf8_byte_sequence(0x800, 2),
+            generate_utf8_byte_sequence(0x1'0000, 3)
+    );
+
+    // Test incomplete continuation bytes
+    auto const it_begin{valid_utf8_byte_sequence.cbegin()};
+    std::string const valid{"Valid"};
+    for (auto it_end{valid_utf8_byte_sequence.cend() - 1};
```

Comment:
We've been using `_it` as a suffix rather than `it_` as a prefix. So can we switch to `begin_it` and `end_it`?